### PR TITLE
[lang] Experimental support of conjugate gradient solver

### PR DESCRIPTION
**Pull request by FantasyVR**
Issue: #7034

### CODE REVIEW

File: python/taichi/linalg/sparse_matrix.py [hunk 1]
```diff
@@ -254,10 +255,10 @@ def build(self, dtype=f32, _format='CSR'):
         taichi_arch = get_runtime().prog.config().arch
         if taichi_arch == _ti_core.Arch.x64 or taichi_arch == _ti_core.Arch.arm64:
             sm = self.ptr.build()
-            return SparseMatrix(sm=sm)
+            return SparseMatrix(sm=sm, dtype=dtype)
         if taichi_arch == _ti_core.Arch.cuda:
             sm = self.ptr.build_cuda()
-            return SparseMatrix(sm=sm)
+            return SparseMatrix(sm=sm, dtype=dtype)
```
Comment (houkensjtu) [inline python/taichi/linalg/sparse_matrix.py:261]: Can we give an error message here for `ti.f64` dtype on CUDA backend?

File: tests/python/test_cg.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import numpy as np
+import pytest
+
+import taichi as ti
+from tests import test_utils
+
+
+@pytest.mark.parametrize("ti_dtype", [ti.f32, ti.f64])
+@test_utils.test(arch=ti.cpu)
```
Comment (houkensjtu) [inline tests/python/test_cg.py:9]: Shall we include a test for CUDA backend here, too?

Comment (FantasyVR) [reply tests/python/test_cg.py:9]: CUDA backend CG solver hasn't been implemented yet.

File: python/taichi/linalg/sparse_matrix.py (hunk 1 above)
Comment (houkensjtu) [reply python/taichi/linalg/sparse_matrix.py:264]: It's probably better to raise the error before doing `sm = self.ptr.build_cuda()`.

**Review by houkensjtu — APPROVED** [review]
Thank you and LGTM!

File: taichi/program/conjugate_gradient.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  if (!CUBLASDriver::get_instance().is_loaded()) {
+    bool load_success = CUBLASDriver::get_instance().load_cublas();
+    if (!load_success) {
+      TI_ERROR("Failed to load cublas library!");
+    }
+  }
+  CUBLASDriver::get_instance().cubCreate(&handle);
+  // int version;
+  // CUBLASDriver::get_instance().cubGetVersion(handle, &version);
+  // fmt::print("CUBLAS version: {}\n", version);
```
Comment (strongoier) [inline taichi/program/conjugate_gradient.cpp:15]: Are we intended to keep these comments?

Comment (FantasyVR) [reply taichi/program/conjugate_gradient.cpp:15]: I'm not sure if we should print out the version of CuBlas.

Comment (strongoier) [reply taichi/program/conjugate_gradient.cpp:15]: We can print info like this using `TI_TRACE`.

**Review by strongoier — APPROVED** [review]

**Pull request merged**